src/main/java/io/appium/java_client/remote/AppiumProtocolHandShake.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.appium.java_client.remote;
+
+import static com.google.common.base.Charsets.UTF_8;
+import static com.google.common.net.HttpHeaders.CONTENT_LENGTH;
+import static com.google.common.net.HttpHeaders.CONTENT_TYPE;
+import static com.google.common.net.MediaType.JSON_UTF_8;
+import static org.openqa.selenium.remote.ErrorCodes.SESSION_NOT_CREATED;
+import static org.openqa.selenium.remote.ErrorCodes.SUCCESS;
+
+import com.google.common.base.Preconditions;
+
+import org.openqa.selenium.Capabilities;
+import org.openqa.selenium.SessionNotCreatedException;
+import org.openqa.selenium.remote.BeanToJsonConverter;
+import org.openqa.selenium.remote.Command;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.Dialect;
+import org.openqa.selenium.remote.ErrorHandler;
+import org.openqa.selenium.remote.JsonException;
+import org.openqa.selenium.remote.JsonToBeanConverter;
+import org.openqa.selenium.remote.Response;
+import org.openqa.selenium.remote.SessionId;
+import org.openqa.selenium.remote.http.HttpClient;
+import org.openqa.selenium.remote.http.HttpMethod;
+import org.openqa.selenium.remote.http.HttpRequest;
+import org.openqa.selenium.remote.http.HttpResponse;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+class AppiumProtocolHandShake {
+
+    public Result createSession(HttpClient client, Command command)
+            throws IOException {
+
+        Capabilities desired = (Capabilities) command.getParameters().get("desiredCapabilities");
+        desired = desired == null ? new DesiredCapabilities() : desired;
+        Capabilities required = (Capabilities) command.getParameters().get("requiredCapabilities");
+        required = required == null ? new DesiredCapabilities() : required;
+
+        String des = new BeanToJsonConverter().convert(desired);
+        String req = new BeanToJsonConverter().convert(required);
+
+        // Assume the remote end obeys the robustness principle.
+        StringBuilder parameters = new StringBuilder("{");
```

Comment:
I'm thinking about using JSONObject here and JSONObject.toString() instead of StringBuilder, which can be potentially not so secure and stable.